### PR TITLE
SE: Learn ObjectConstraint on branching from constant pattern

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Contributes to #5970

Rebase will be needed, commit `Squash Equals` doesn't need a review.

### CODE REVIEW

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]
The comment should be updated.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Pattern.cs [hunk 1]
```diff
@@ -32,15 +32,44 @@ internal static class Pattern
...
+            {
+                OperationKindEx.ConstantPattern => ConstraintFromConstantPattern(state, IConstantPatternOperationWrapper.FromOperation(pattern.WrappedOperation), useOpposite),
+                _ => null
+            };
+
+        private static SymbolicConstraint ConstraintFromConstantPattern(ProgramState state, IConstantPatternOperationWrapper constant, bool useOpposite)
+        {
+            if (state[constant.Value] is { } value)
+            {
+                if (value.Constraint<BoolConstraint>() is { } boolConstraint && !useOpposite)   // Cannot use opposite on booleans. If it is not "true", it could be null, false or any other type
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Pattern.cs:61]: `is true` and `is false` is not practically needed for boolean types. The comment explains what happens with the rest.

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs [hunk 2]
```diff
@@ -302,5 +302,120 @@ End If
...
+{{
+    Tag(""Else"", arg);
+}}
+Tag(""End"", arg);";
+            var validator = SETestContext.CreateCS(code, ", object arg").Validator;
+            validator.ValidateTag("If", x => x.HasConstraint(ObjectConstraint.Null).Should().BeTrue());
+            validator.ValidateTag("Else", x => x.HasConstraint(ObjectConstraint.NotNull).Should().BeTrue());
+            validator.TagValues("End").Should().HaveCount(2)
+                .And.ContainSingle(x => x.HasConstraint(ObjectConstraint.Null))
+                .And.ContainSingle(x => x.HasConstraint(ObjectConstraint.NotNull));
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs:326]: We'll hopefully improve the assertion framework

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs [hunk 3]
```diff
@@ -302,5 +302,120 @@ End If
...
+            var code = @$"
+if ({expression})
+{{
+    Tag(""If"", arg);
+}}
+else
+{{
+    Tag(""Else"", arg);
+}}
+Tag(""End"", arg);";
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs:320]: While it branches, it makes the branching flow less obvious with the return value

**Review by martin-strecker-sonarsource — APPROVED** [review]
LGTM

**Pull request merged**